test/Microsoft.AspNetCore.Mvc.Core.Test/DependencyInjection/ApplicationModelConventionExtensionsTest.cs

```diff
@@ -182,5 +264,37 @@ public void Apply(ControllerModel controller)
                 controller.Properties.Add("TestProperty", "TestValue");
             }
         }
+
+        private class ControllerModelCollectionModifyingConvention : IControllerModelConvention
+        {
+            public void Apply(ControllerModel controller)
+            {
+                controller.Application.Controllers.Remove(controller);
+            }
+        }
+
+        private class TestApplicationModelConvention : IApplicationModelConvention
+        {
+            public void Apply(ApplicationModel application)
+            {
+                application.Controllers.RemoveAt(0);
+            }
+        }
+
+        private class ActionModelCollectionModifyingConvention : IActionModelConvention
+        {
+            public void Apply(ActionModel action)
+            {
+                action.Controller.Actions.Remove(action);
```

Comment:
Does this work well if you add an action? We could possibly support it (if we indexed in to the collection and tracked if it's still there to track removals). Would be good to find out more from @rynowak what the exact scenario was.

Reply:
Nothing complicated is needed here. The user code sees the original list.